src/dune_engine/build_system.ml

```diff
@@ -1517,6 +1543,9 @@ end = struct
             match sandbox with
             | None -> Path.Build.Set.empty
             | Some sandbox ->
+              (* The stamp file for anonymous actions is always created outside
+                 the sandbox, so we can't move it. *)
+              let targets = targets_without_stamp_file targets rule_kind in
```

Comment:
Agreed. I will do the change. BTW, would we need both `Targets.t` and `Targets.Validated.t`?

Reply:
Yes, I thought that initially, when we are constructing the set of targets, we would use `Targets.t`. But later, calling `Targets.validate` would return `Targets.Validated.t`, so `Rule.t` would have `targets : Targets.Validated.t`. In this way, we would have more static guarantees. 

I already have a local branch doing these changes, was just distracted from finishing it. If you think, it's a good idea, I'll submit a PR later today.

Reply:
That seems good to me.

Reply:
So I can't quite detect which targets are part of the sandbox as `Sandbox.move_targets_to_build_dir` takes non-sandboxed path and systematically sandbox them. I could change the function to take a set of files and a set of dirs instead of a `Targets.t`. WDYT?

Reply:
Ah, I see. Yes, then passing a set of files and directories to move out of the sandbox seems fine.